server: prevent test args race in shared-process multi-node cluster

yuzefovich opened the pull request:
This commit prevents a possible race that could previously occur between the shared-process tenant being started explicitly with the correct args (via `Activate` call) and server controller realizing that a tenant is missing and starting it with no test args. It is achieved by setting the correct test args into the controller for the default shared-process tenant in server's `PreStart` method. All servers are pre-started before any is activated, so now even if the default tenant is started via the server controller on one of the nodes, it'll have the correct test args set.

Fixes: #113598.
Fixes: #113854.
Fixes: #113855.
Fixes: #113856.
Fixes: #113857.
Fixes: #113641.
Fixes: #113700.
Fixes: #113582.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114455)
<!-- Reviewable:end -->

yuzefovich committed:
server: prevent test args race in shared-process multi-node cluster

This commit prevents a possible race that could previously occur between
the shared-process tenant being started explicitly with the correct args
(via `Activate` call) and server controller realizing that a tenant is
missing and starting it with no test args. It is achieved by setting the
correct test args into the controller for the default shared-process
tenant in server's `PreStart` method. All servers are pre-started before
any is activated, so now even if the default tenant is started via the
server controller on one of the nodes, it'll have the correct test args
set.

Release note: None

yuzefovich commented:
Ricky [thinks](https://cockroachlabs.slack.com/archives/C023S0V4YEB/p1700007772339909) that failures under stress / stressrace are expected given known deficiencies how CI does `stress`. Perhaps shared-process multi-tenant has the largest overhead of 3 modes, so it's more likely to hit those.

maryliag reviewed (approved):
Nice!! Thank you for working on this fix!
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114455#-:-NjIMSJb0g8ztV22sQon:b-pwcu8e)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan and @stevendanna)



<!-- Sent from Reviewable.io -->

maryliag commented:
Btw, I think this will also solve the initial PRs that Herko used to create the issues you tagged, so can you also tag as fixing:

24.1:
https://github.com/cockroachdb/cockroach/issues/113641
https://github.com/cockroachdb/cockroach/issues/113700

When this is backported to 23.2:
https://github.com/cockroachdb/cockroach/issues/113783
https://github.com/cockroachdb/cockroach/issues/113785
https://github.com/cockroachdb/cockroach/issues/113828
https://github.com/cockroachdb/cockroach/issues/114328

Thank you!

herkolategan reviewed (approved):
Nice catch! Thanks for fixing. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114455#-:-NjIoEqn9q6uwJURU91e:b-kc6cxp)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @stevendanna)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
